packages/store/src/database/backends/indexedDB.ts

```diff
@@ -1,48 +1,322 @@
+      baseQuery = compareUsingPredicate(predicatedQuery, query.predicate);
+
+      let includedTaskIds: TaskID[];
+      // only one possible key right now, when another key is eventually defined convert this
+      // to be an if-else
+      switch (query.predicate[0]) {
+        case "dueTimestampMillis":
+          const clause = this.db.derived_taskComponents.where(
+            DexieDerivedTaskComponentColumn.DueTimestampMillis,
+          );
+          const derivedRowsPrimaryKeys = await compareUsingPredicate(
+            clause,
+            query.predicate,
+          ).primaryKeys();
+
+          includedTaskIds = derivedRowsPrimaryKeys.map(
+            ([taskID]) => taskID as TaskID,
+          );
+      }
+
+      // was the afterID specified at the same time?
+      if (query.afterID) {
+        // simulate a join across the two tables
+        const afterRowID = await this._fetchPrimaryKeyFromUniqueKey(
+          this.db.entities,
+          DexieEventColumn.ID,
+          query.afterID,
+        );
+
+        baseQuery = this.db.entities
+          .where(DexieEntityColumn.RowID)
+          .above(afterRowID)
+          .filter((row) => includedTaskIds.includes(row.id));
+      } else {
+        baseQuery = this.db.entities
+          .where(DexieEntityColumn.ID)
+          .anyOf(includedTaskIds);
+      }
+    } else if (query.afterID) {
+      const afterRowID = await this._fetchPrimaryKeyFromUniqueKey(
+        this.db.entities,
+        DexieEventColumn.ID,
+        query.afterID,
+      );
+      baseQuery = this.db.entities
+        .where(DexieEntityColumn.RowID)
+        .above(afterRowID);
+    } else {
+      baseQuery = this.db.entities.orderBy(DexieEntityColumn.RowID);
+    }
+
+    if (query.limit) {
+      baseQuery = baseQuery.limit(query.limit);
+    }
+
+    const queriedEntities = await baseQuery.toArray();
+    return queriedEntities.map((entity) => ({
+      lastEventID: entity.lastEventID,
+      entity: JSON.parse(entity.data),
+    }));
   }
 
-  listEvents(query: DatabaseEventQuery): Promise<Event[]> {
-    return Promise.resolve([]);
+  async listEvents(query: DatabaseEventQuery): Promise<Event[]> {
+    let baseQuery: Dexie.Collection<DexieEventRow, number>;
+    if (query.predicate && query.afterID) {
+      const afterSequenceNumber = await this._fetchPrimaryKeyFromUniqueKey(
+        this.db.events,
+        DexieEventColumn.ID,
+        query.afterID,
+      );
+
+      // both the predicate and afterID were specificed need to do more complex querying
+      if (query.predicate[1] == "=") {
+        // can create a compound query in this case
+        baseQuery = this.db.events
+          .where(`[${query.predicate[0]}+${DexieEventColumn.SequenceNumber}]`)
+          .between(
+            [query.predicate[2], afterSequenceNumber + 1],
+            [query.predicate[2], Dexie.maxKey],
+          );
+      } else {
+        // arbitrary comparsion query
+        baseQuery = this.db.events.filter(
```

Comment:
> Alternately (in the future—don't need to worry about this now... though maybe a comment?), we could pursue the same strategy you did for dueTimestamp: use an index to find event IDs relevant to the predicate, seek the primary key table to the row ID corresponding to the afterID, and filter by the event IDs found in the first step. This'll only require scanning all values matching the predicate, rather than the full table.

Just realized that what I was saying here is nonsense. The strategy I'm describing would still involve scanning all events after the one corresponding to `afterID`. But now I see we can do a lot better:

```
const predicatedQuery = this.db.events.where(query.predicate[0]);
const collection = compareUsingPredicate(predicatedQuery, query.predicate)
  .filter((item) => item[DexieEventColumn.SequenceNumber] > afterSequenceNumber)
  .sortBy(DexieEventColumn.SequenceNumber);
```

This approach avoids a full table scan altogether. It does still wastefully scan all events matching the predicate (rather than just those needed for the current page), but it at least doesn't need to keep in memory and sort those we've paged past.

Reviewing the performance implications of paged iteration (let N be the count of events and M be the count of events matching the predicate):

A. The original approach in your PR: time cost `O(N^2)`; space cost `O(1)`

B. The first approach described in my previous comment (seek table by sequence number then filter by predicate): time cost `O(N^2)` (but constant roughly halved compared to A); space cost `O(1)`

C. The second approach described in my previous comment (query index by predicate; seek table by sequence number then filter by IDs): time cost `O(N^2)` (constant similar to B); space cost `O(M)`. This is just worse than B.

D. The approach described in this comment (query index by predicate; filter by sequence number and sort): time cost `O(M * log N)`; space cost `O(M)`.

The dominant strategies here are B and D. D has a better time cost but a worse space cost. It'll be the right choice so long as `M` is not too large. We may or may not be able to know that in advance, of course.

Reply:
Thank you for all that analysis. Super helpful. Do we need to sort in this case? Wouldn't the rows already be sorted?

Reply:
This thread disappeared in the latest changes, so see reply inline [here](https://github.com/andymatuschak/orbit/pull/225/files/fa8e84a44b9e6c3837295ce0fa9a98a5965fedf2#diff-ca29974e99a98a9c3e1292ab3970c65eee5bd8b269aabb6be5b7ea23e2b22288).